near-sdk/src/json_types/public_key.rs

```diff
@@ -104,10 +119,13 @@ impl<'de> serde::Deserialize<'de> for Base58PublicKey {
 
 impl From<&Base58PublicKey> for String {
     fn from(str_public_key: &Base58PublicKey) -> Self {
-        match str_public_key.0[0] {
-            0 => "ed25519:".to_string() + &bs58::encode(&str_public_key.0[1..]).into_string(),
-            1 => "secp256k1:".to_string() + &bs58::encode(&str_public_key.0[1..]).into_string(),
-            _ => panic!("Unexpected curve"),
+        match str_public_key.curve {
+            CurveType::ED25519 => {
+                "ed25519:".to_string() + &bs58::encode(&str_public_key.data).into_string()
+            }
+            CurveType::SECP256K1 => {
+                "secp256k1:".to_string() + &bs58::encode(&str_public_key.data).into_string()
+            }
```

Comment:
```suggestion
        match str_public_key.curve {
            CurveType::ED25519 => {
                ["ed25519:", &bs58::encode(&str_public_key.data).into_string()].concat()
            }
            CurveType::SECP256K1 => {
                ["secp256k1:", &bs58::encode(&str_public_key.data).into_string()].concat()
            }
        }
```

This one is a bit optional because I'm actually not sure about the internals if it can avoid the reallocation on adding the other string (since you're not specifying a capacity) and I think concat does that, but I could be wrong.

Reply:
welp, I went down the rabbithole for this one. `str.to_string()` allocates a `string` with a capacity, so the `+` operator will indeed do a second allocation. The concat version is better here